### ISSUE TITLE
Use move in variant_deserializer

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3734]
The current implementation:
```C++
template <cluster_version_t W, int N, class Variant, class T, class... Ts>
struct variant_deserializer<W, N, Variant, T, Ts...> {
    static MUST_USE archive_result_t deserialize_variant(int n, read_stream_t *s, Variant *x) {
        if (n == N) {
            T v;
            archive_result_t res = deserialize<W>(s, &v);
            if (bad(res)) { return res; }
            *x = v;

            return archive_result_t::SUCCESS;
        } else {
            return variant_deserializer<W, N + 1, Variant, Ts...>::deserialize_variant(n, s, x);
        }
    }
};
```

We should make this use
```C++
*x = std::move(v);
```

to avoid a potentially expensive copy of the variant (some variants can be vectors for example).

**Comment by danielmewes** [comment]
(I'm not sure if `std::move` works with boost variants like that, but hopefully something similar to that works)

**Comment by mlucy** [comment]
I think some of the boost versions we compile with don't support move semantics.  We could instead do `*x = T(); deserialize<W>(s, boost::get<T>(x))` or something.

**Comment by danielmewes** [comment]
Fixed in `next` 091190627e65e72e8343208d5b8be12326945077
